pkg/agent/openflow/pipeline.go

```diff
@@ -1410,163 +1455,90 @@ func getIPProtocol(ip net.IP) binding.Protocol {
 	var flows []binding.Flow
-	for _, proto := range c.ipProtocols {
-		nextTable := SpoofGuardTable.GetNext()
-		if proto == binding.ProtocolIPv6 {
-			nextTable = IPv6Table.GetID()
+	for _, ipProtocol := range f.ipProtocols {
+		targetTable := SpoofGuardTable.ofTable.GetNext()
+		if ipProtocol == binding.ProtocolIPv6 {
+			targetTable = IPv6Table.ofTable.GetID()
 		}
 		flows = append(flows,
-			SpoofGuardTable.BuildFlow(priorityNormal).MatchProtocol(proto).
+			SpoofGuardTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchProtocol(ipProtocol).
 				MatchInPort(config.HostGatewayOFPort).
-				Action().GotoTable(nextTable).
-				Cookie(c.cookieAllocator.Request(category).Raw()).
+				Action().GotoTable(targetTable).
 				Done(),
 		)
 	}
 	return flows
 }
 
-// serviceCIDRDNATFlow generates flows to match dst IP in service CIDR and output to host gateway interface directly.
-func (c *client) serviceCIDRDNATFlows(serviceCIDRs []*net.IPNet) []binding.Flow {
+// serviceCIDRDNATFlows generates the flows to match destination IP in Service CIDR and output to the Antrea gateway directly.
+func (f *featureService) serviceCIDRDNATFlows(serviceCIDRs []*net.IPNet) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
 	var flows []binding.Flow
 	for _, serviceCIDR := range serviceCIDRs {
 		if serviceCIDR != nil {
-			ipProto := getIPProtocol(serviceCIDR.IP)
-			flows = append(flows, DNATTable.BuildFlow(priorityNormal).MatchProtocol(ipProto).
+			ipProtocol := getIPProtocol(serviceCIDR.IP)
+			flows = append(flows, DNATTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchProtocol(ipProtocol).
 				MatchDstIPNet(*serviceCIDR).
 				Action().LoadToRegField(TargetOFPortField, config.HostGatewayOFPort).
 				Action().LoadRegMark(OFPortFoundRegMark).
-				Action().GotoTable(ConntrackCommitTable.GetID()).
-				Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+				Action().GotoStage(stageConntrack).
 				Done())
 		}
 	}
 	return flows
 }
 
-// serviceNeedLBFlow generates flows to mark packets as LB needed.
-func (c *client) serviceNeedLBFlow() binding.Flow {
-	return SessionAffinityTable.BuildFlow(priorityMiss).
-		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+// serviceNeedLBFlow generates the default flow to mark packets with EpToSelectRegMark.
+func (f *featureService) serviceNeedLBFlow() binding.Flow {
+	return SessionAffinityTable.ofTable.BuildFlow(priorityMiss).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
 		Action().LoadRegMark(EpToSelectRegMark).
 		Done()
 }
 
-// arpResponderLocalFlows generates the ARP responder flow entry that replies request from local Pods for local
-// gateway MAC.
-// Only used in AntreaFlexibleIPAM to avoid multiple ARP replies from antrea-gw0 and uplink.
-// TODO(gran): use better method to process ARP and support IPv6.
-func (c *client) arpResponderLocalFlows(category cookie.Category) (flows []binding.Flow) {
```

Comment:
Is this removed? How the above case is handled now?

Reply:
It is replace by the following function:
```go
// arpResponderFlow generates the flow to reply to the ARP request with a MAC address for the target IP address.
func (f *featurePodConnectivity) arpResponderFlow(ipAddr net.IP, macAddr net.HardwareAddr) binding.Flow {
	return ARPResponderTable.ofTable.BuildFlow(priorityNormal).
		Cookie(f.cookieAllocator.Request(f.category).Raw()).
		MatchProtocol(binding.ProtocolARP).
		MatchARPOp(arpOpRequest).
		MatchARPTpa(ipAddr).
		Action().Move(binding.NxmFieldSrcMAC, binding.NxmFieldDstMAC).
		Action().SetSrcMAC(macAddr).
		Action().LoadARPOperation(arpOpReply).
		Action().Move(binding.NxmFieldARPSha, binding.NxmFieldARPTha).
		Action().SetARPSha(macAddr).
		Action().Move(binding.NxmFieldARPSpa, binding.NxmFieldARPTpa).
		Action().SetARPSpa(ipAddr).
		Action().OutputInPort().
		Done()
}
```

Reply:
Yes,  it should be installed when connectUplinkToBridge is enabled.

Reply:
But it seems now it's installed even when `connectUplinkToBridge` is disabled. Is it expected?